ftprime/argrecorder.py

```diff
@@ -1,48 +1,117 @@
+
+    The internal state is stored using
+        - self.node_ids[k] : gives the output Node ID corresponding to the
+          input individual ID k+self.input_min
+    and self.birth_times and self.populations stores the corresponding birth
+    times (in forwards-time) and populations.  These need to contain information
+    about every input individual who might be seen in a new edgeset (ie, who
+    might be a parent).
+
+    Assumes that input individual IDs are *nondecreasing*.
+
+    Must be initialized with a tree sequence ``ts``, this will serve as
+    the history of this first generation of individuals.  The ``samples`` of
+    this tree sequence correspond to the individuals in the first generation.
+
+    If different than the sample IDs of ``ts``, ``first_generation`` has the
+    node IDs of the first generation.
     '''
 
-    def __init__(self):
-        super(ARGrecorder, self).__init__()
-        # need to keep track of this to know where to append new individuals
-        # and how many rows in the NodeTable there are
-        self.num_nodes = 0
-        self.max_time = 0
+    def __init__(self, ts, first_generation=ts.samples(), time=0.0, max_indivs=1000):
+        """
+        :param ts TreeSequence: The tree sequence defining relationships between
+            the first genreation of individuals.
+        :param list first_generation: The input IDs of the first generation that
+            should be mapped onto the samples in the tree sequence ``ts``.
+        :param float time: The (forwards) time of the "present" at the start of
+            the simulation.
+        :param int max_indivs: The largest number of individuals (past or present)
+            that we will track at any given time.
+        """
+        # maximum number of individuals that can be kept track of in the internal state
+        self.n_max = max_indivs
+        # this is output node ID of the next new node
+        self.num_nodes = ts.num_nodes # n
+        # this is the largest (forwards) time seen so far
+        self.max_time = time  # T
+        # last (forwards) time we updated node times
+        self.last_update_time = time  # T_0
+        # number of nodes that have the time right
+        self.last_update_node = ts.num_nodes
+        assert len(ts.samples()) == len(first_generation)
+        # minimum currently active input ID
+        self.input_min = 0  # m_0
+        # list of output node IDs indexed by input labels
+        self.node_ids = np.empty(self.n_max, dtype='int32') # L
+        self.reset_node_ids(first_generation, ts.samples())
+        # the actual tables that get updated
+        #  DON'T actually store ts, just the tables:
+        self.nodes = msprime.NodeTable()
+        self.edgesets = msprime.EdgesetTable()
+        self.sites = msprime.SiteTable()
+        self.mutations = msprime.MutationTable()
+        self.tables = ts.dump_tables(nodes=self.nodes, edgesets=self.edgesets,
```

Comment:
Are you using ``self.tables`` here somewhere? It seems a little bit confusing to pass the tables as arguments, and also use the return value. I had imagined the return value as a convenience for when you don't want to bother with allocating tables yourself, i.e.
```
self.tables = ts.dump_tables()
```